### PR TITLE
[axs15231] Touchscreen driver

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

Add support
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4334

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by xiasi0** [comment]
St77916 also uses QSPI protocol. We hope you can add it. Thank you 🙏

**Comment by clydebarrow** [comment]
> St77916 also uses QSPI protocol. We hope you can add it. Thank you 🙏

See https://github.com/esphome/esphome-docs/pull/4335 - the CUSTOM model with an appropriate init sequence should work.

File: esphome/components/axs15231/touchscreen/axs15231_touchscreen.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+void AXS15231Touchscreen::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up AXS15231 Touchscreen...");
+  if (this->reset_pin_ != nullptr) {
+    this->reset_pin_->setup();
+    this->reset_pin_->digital_write(false);
+    delay(5);
+    this->reset_pin_->digital_write(true);
+    delay(10);  // NOLINT
```
Comment (jesserockz) [inline esphome/components/axs15231/touchscreen/axs15231_touchscreen.cpp]: Not needed for only 10ms
```suggestion
    delay(10);
```

File: esphome/components/axs15231/touchscreen/axs15231_touchscreen.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  err = this->write(AXS_READ_TOUCHPAD, sizeof(AXS_READ_TOUCHPAD), false);
+  ERROR_CHECK(err);
+  err = this->read(data, sizeof(data));
+  ERROR_CHECK(err);
+  if (data[0] != 0)  // no touches
+    return;
+  uint16_t x = encode_uint16(data[2] & 0xF, data[3]);
+  uint16_t y = encode_uint16(data[4] & 0xF, data[5]);
+  this->add_raw_touch_position_(0, x, y);
+}
```
Comment (jesserockz) [inline esphome/components/axs15231/touchscreen/axs15231_touchscreen.cpp:52]: Don't forget to clear previous warnings if it made it this far.
```suggestion
  this->status_clear_warning();
}
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
A Couple of minor suggestions below, otherwise LGTM

**Review by jesserockz — APPROVED** [review]

**Pull request merged**